odl/deform/mass_preserving.py

```diff
@@ -0,0 +1,548 @@
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        if self.domain == self.deformation.space[0]:
+            domain_repr = ''
+        else:
+            domain_repr = ', domain={!r}'.format(self.domain)
+
+        return '{}({!r}{})'.format(self.__class__.__name__,
+                                   self.deformation,
+                                   domain_repr)
+
+
+class MassPresvDeformFixedTempl(Operator):
+
+    """Mass-preserving deformation operator with fixed template.
+
+    The operator has a fixed template ``I`` and maps a mass-preserving
+    deformation field, i.e., a deformation field ``phi`` and its Jacobian
+    determinant to the new function ``x --> |Dphi(x)|I(phi(x))``.
+
+
+    See Also
+    --------
+    MassPresvDeformFixedDeform : Mass-preserving deformation with fixed
+    mass-preserving deformation field.
+
+    Notes
+    -----
+    For :math:`\Omega \subset \mathbb{R}^d`, we take :math:`X = L^p(\Omega)`
+    to be the template space, i.e. :math:`I \\in X`. Then the deformation field
+    space is identified with :math:`G := X^d`. Hence the deformation operator
+    with fixed template maps :math:`G` into :math:`X`:
+
+    .. math::
+
+        W_I : G \\to X, \quad W_I(\phi) := |D\phi(\cdot)|I(\phi(\cdot)),
+
+    i.e., :math:`W_I(\phi)(x) = |D\phi(x)|I(\phi(x))`.
+    """
+
+    def __init__(self, template):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template that is to be deformed.
+
+        Examples
+        --------
+        Create a simple 1D template ``I`` over ``[0, 1]`` to initialize the
+        operator and apply it to a mass-preserving deformation field, i.e.,
+        a deffeomorphic deformation field and its Jacobian determinant.
+        If the deffeomorphism ``phi`` is ``sin(x*pi/2)``, its Jacobian
+        determinant should be ``cos(x*pi/2)*pi/2`` and the output value is
+        ``cos(x*pi/2)*pi/2*I(sin(x*pi/2))``. A mass-preserving deformation
+        is obtained, namely, the integration of ``I(x)`` over ``[0, 1]``
+        equals to the integration of ``|Dphi(x)|I(phi(x))`` over ``[0, 1]``.
+
+        >>> import odl
+        >>> import numpy as np
+        >>> space = odl.uniform_discr(0, 1, 5, interp='linear')
+        >>> vspace = space.vector_field_space
+        >>> pspace = ProductSpace(vspace, space)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> op = MassPresvDeformFixedTempl(template)
+        >>> img_pts = space.points()
+        >>> deform_field = vspace.element([np.sin(img_pts * np.pi / 2.0)])
+        >>> det = space.element(np.cos(img_pts * np.pi / 2.0) * np.pi / 2.0)
+        >>> mass_presv_deform = pspace.element()
+        >>> mass_presv_deform[0] = deform_field
+        >>> mass_presv_deform[1] = det
+        >>> print(op(mass_presv_deform))
+        [0.0, 1.07761764468, 0.0, 0.0, 0.0]
```

Comment:
This result is not obvious from the text and code, possibly because the example is quite length. Some further explanation could be needed.